File: data_db/src/main/java/br/org/restaurantepopular/data_db/repository/category/CategoryRepository.java
package br.org.restaurantepopular.data_db.repository.category;

import br.org.restaurantepopular.data_db.mapper.CategoryEntityMapper;
import br.org.restaurantepopular.entity.Category;
import br.org.restaurantepopular.usecase.category.ICategoryRepository;
import br.org.restaurantepopular.usecase.core.exception.NotFoundException;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class CategoryRepository implements ICategoryRepository {

    private final CategoryDAO categoryDAO;

    public CategoryRepository(CategoryDAO categoryDAO) {
        this.categoryDAO = categoryDAO;
    }

    @Override
    public Optional<Category> save(Category category) {
        return Optional.ofNullable(CategoryEntityMapper.toCategory(
                categoryDAO.save(CategoryEntityMapper.toCategoryEntity(category))
        ));
    }

    @Override
    public void update(Category category) throws NotFoundException {
        if (!categoryDAO.existsById(category.getId()))
            throw new NotFoundException(String.format("Not found a category with this %s ID", category.getId()));

        categoryDAO.save(CategoryEntityMapper.toCategoryEntity(category));
    }

    @Override
    public List<Category> listCategories() {
        return categoryDAO
                .findAll()
                .stream()
                .map(CategoryEntityMapper::toCategory)
                .collect(Collectors.toList());
    }

    @Override
    public Optional<Category> findById(String id) throws NotFoundException {
        var categoryOptional = categoryDAO.findById(id).map(CategoryEntityMapper::toCategory);
        if (categoryOptional.isEmpty())
            throw new NotFoundException(String.format("Not found a category with this %s ID", id));
        return categoryOptional;
    }

    @Override
    public void delete(String id) {
        if (!categoryDAO.existsById(id))
            throw new NotFoundException(String.format("Not found a category with this %s ID", id));
        categoryDAO.deleteById(id);
    }
}

File: data_db/src/main/java/br/org/restaurantepopular/data_db/repository/category/CategoryRepository.kt
package br.org.restaurantepopular.data_db.repository.category

import br.org.restaurantepopular.data_db.entity.CategoryEntity
import br.org.restaurantepopular.data_db.mapper.toCategory
import br.org.restaurantepopular.data_db.mapper.toCategoryEntity
import br.org.restaurantepopular.entity.Category
import br.org.restaurantepopular.usecase.category.ICategoryRepository
import br.org.restaurantepopular.usecase.core.exception.NotFoundException

class CategoryRepository(
    private val categoryDAO: CategoryDAO
) : ICategoryRepository {

    override fun save(category: Category): Category {
        return categoryDAO.save(category.toCategoryEntity()).toCategory()
    }

    @Throws(NotFoundException::class)
    override fun update(category: Category) {
        if (!categoryDAO.existsById(category.id!!))
            throw NotFoundException(String.format("Not found a category with this %s ID", category.id))

        categoryDAO.save(category.toCategoryEntity())
    }

    override fun listCategories(): List<Category> {
        return categoryDAO
            .findAll()
            .map(CategoryEntity::toCategory)
    }

    @Throws(NotFoundException::class)
    override fun findById(id: String): Category {
        val categoryOptional =
            categoryDAO.findById(id)
                .map(CategoryEntity::toCategory)

        if (categoryOptional.isEmpty)
            throw NotFoundException(String.format("Not found a category with this %s ID", id))
        return categoryOptional.get()
    }

    override fun delete(id: String) {
        if (!categoryDAO.existsById(id))
            throw NotFoundException(String.format("Not found a category with this %s ID", id))

        categoryDAO.deleteById(id)
    }
}
